File: server/src/test/java/com/larsreimann/api_editor/server/file_handling/ModuleStubContentBuilderTest.java
package com.larsreimann.api_editor.server.file_handling;

import com.larsreimann.api_editor.server.data.*;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;

class ModuleStubContentBuilderTest {
    @Test
    void buildModuleContentReturnsFormattedModuleContent() {
        // given
        AnnotatedPythonClass testClass = new AnnotatedPythonClass(
            "test-class",
            "test-module.test-class",
            List.of("test-decorator"),
            List.of("test-superclass"),
            List.of(
                new AnnotatedPythonFunction(
                    "test-class-function",
                    "test-module.test-class.test-class-function",
                    List.of("decorators"),
                    List.of(
                        new AnnotatedPythonParameter(
                            "only-param",
                            "test-module.test-class.test-class-function.only-param",
                            "'defaultValue'",
                            PythonParameterAssignment.POSITION_OR_NAME,
                            true,
                            "typeInDocs",
                            "description",
                            Collections.emptyList()
                        )),
                    Collections.emptyList(),
                    true,
                    "description",
                    "fullDocstring",
                    Collections.emptyList()
                ),
                new AnnotatedPythonFunction(
                    "__init__",
                    "test-module.test-class.__init__",
                    List.of("decorators"),
                    List.of(new AnnotatedPythonParameter(
                        "only-param",
                        "test-module.test-class.__init__.only-param",
                        "'defaultValue'",
                        PythonParameterAssignment.POSITION_OR_NAME,
                        true,
                        "typeInDocs",
                        "description",
                        Collections.emptyList()
                    )),
                    Collections.emptyList(),
                    true,
                    "description",
                    "fullDocstring",
                    Collections.emptyList()
                )),
            "Lorem ipsum",
            "Lorem ipsum",
            Collections.emptyList()
        );

        AnnotatedPythonModule testModule = new AnnotatedPythonModule(
            "test-module",
            Collections.emptyList(),
            Collections.emptyList(),
            List.of(
                testClass
            ),
            List.of(
                new AnnotatedPythonFunction(
                    "function_module",
                    "test-module.function_module",
                    List.of("test-decorator"),
                    List.of(
                        new AnnotatedPythonParameter(
                            "param1",
                            "test-module.function_module.param1",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        ),
                        new AnnotatedPythonParameter(
                            "param2",
                            "test-module.function_module.param2",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        ),
                        new AnnotatedPythonParameter(
                            "param3",
                            "test-module.function_module.param3",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        )
                    ),
                    List.of(
                        new AnnotatedPythonResult(
                            "test-result",
                            "str",
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        )
                    ),
                    true,
                    "Lorem ipsum",
                    "Lorem ipsum",
                    Collections.emptyList()
                ),
                new AnnotatedPythonFunction(
                    "test-function",
                    "test-module.test-function",
                    List.of("test-decorator"),
                    List.of(
                        new AnnotatedPythonParameter(
                            "test-parameter",
                            "test-module.test-function.test-parameter",
                            "42",
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "int",
                            "Lorem ipsum",
                            Collections.emptyList()
                        )
                    ),
                    List.of(
                        new AnnotatedPythonResult(
                            "test-result",
                            "str",
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        )
                    ),
                    true,
                    "Lorem ipsum",
                    "Lorem ipsum",
                    Collections.emptyList()
                )
            ),
            Collections.emptyList()
        );

        // when
        ModuleStubContentBuilder moduleStubContentBuilder =
            new ModuleStubContentBuilder(testModule);
        String moduleContent = moduleStubContentBuilder.buildModuleContent();

        //then
        String expectedModuleContent = """
            package simpleml.test-module

            open class test-class(only-param: Any? or "defaultValue") {
                attr only-param: Any?

                fun test-class-function(only-param: Any? or "defaultValue")
            }

            fun function_module(param1: Any?, param2: Any?, param3: Any?) -> test-result: str

            fun test-function(test-parameter: Any? or 42) -> test-result: str
            """;

        Assertions.assertEquals(expectedModuleContent, moduleContent);
    }

    @Test
    void buildModuleContentWithNoClassesReturnsFormattedModuleContent() {
        // given
        AnnotatedPythonModule testModule = new AnnotatedPythonModule(
            "test-module",
            Collections.emptyList(),
            Collections.emptyList(),
            Collections.emptyList(),
            List.of(
                new AnnotatedPythonFunction(
                    "function_module",
                    "test-module.function_module",
                    List.of("test-decorator"),
                    List.of(
                        new AnnotatedPythonParameter(
                            "param1",
                            "test-module.function_module.param1",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        ),
                        new AnnotatedPythonParameter(
                            "param2",
                            "test-module.function_module.param2",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        ),
                        new AnnotatedPythonParameter(
                            "param3",
                            "test-module.function_module.param3",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        )
                    ),
                    List.of(
                        new AnnotatedPythonResult(
                            "test-result",
                            "str",
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        )
                    ),
                    true,
                    "Lorem ipsum",
                    "Lorem ipsum",
                    Collections.emptyList()
                ),
                new AnnotatedPythonFunction(
                    "test-function",
                    "test-module.test-function",
                    List.of("test-decorator"),
                    List.of(
                        new AnnotatedPythonParameter(
                            "test-parameter",
                            "test-module.test-function.test-parameter",
                            "42",
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "int",
                            "Lorem ipsum",
                            Collections.emptyList()
                        )
                    ),
                    List.of(
                        new AnnotatedPythonResult(
                            "test-result",
                            "str",
                            "str",
                            "Lorem ipsum",
                            Collections.emptyList()
                        )
                    ),
                    true,
                    "Lorem ipsum",
                    "Lorem ipsum",
                    Collections.emptyList()
                )
            ),
            Collections.emptyList()
        );

        // when
        ModuleStubContentBuilder moduleStubContentBuilder =
            new ModuleStubContentBuilder(testModule);
        String moduleContent = moduleStubContentBuilder.buildModuleContent();

        //then
        String expectedModuleContent = """
            package simpleml.test-module

            fun function_module(param1: Any?, param2: Any?, param3: Any?) -> test-result: str

            fun test-function(test-parameter: Any? or 42) -> test-result: str
            """;

        Assertions.assertEquals(expectedModuleContent, moduleContent);
    }

    @Test
    void buildModuleContentWithOnlyConstructorReturnsFormattedModuleContent() {
        // given
        AnnotatedPythonClass testClass = new AnnotatedPythonClass(
            "test-class",
            "test-module.test-class",
            List.of("test-decorator"),
            List.of("test-superclass"),
            List.of(
                new AnnotatedPythonFunction(
                    "__init__",
                    "test-module.test-class.__init__",
                    List.of("decorators"),
                    List.of(new AnnotatedPythonParameter(
                        "only-param",
                        "test-module.test-class.__init__.only-param",
                        "'defaultValue'",
                        PythonParameterAssignment.POSITION_OR_NAME,
                        true,
                        "typeInDocs",
                        "description",
                        Collections.emptyList()
                    )),
                    Collections.emptyList(),
                    true,
                    "description",
                    "fullDocstring",
                    Collections.emptyList()
                )
            ),
            "Lorem ipsum",
            "Lorem ipsum",
            Collections.emptyList()
        );

        AnnotatedPythonModule testModule = new AnnotatedPythonModule(
            "test-module",
            List.of(
                new PythonImport(
                    "test-import1",
                    "test-alias"
                )
            ),
            List.of(
                new PythonFromImport(
                    "test-from-import1",
                    "test-declaration1",
                    null
                )
            ),
            List.of(
                testClass
            ),
            Collections.emptyList(),
            Collections.emptyList()
        );

        // when
        ModuleStubContentBuilder moduleStubContentBuilder =
            new ModuleStubContentBuilder(testModule);
        String moduleContent = moduleStubContentBuilder.buildModuleContent();

        //then
        String expectedModuleContent = """
            package simpleml.test-module

            open class test-class(only-param: Any? or "defaultValue") {
                attr only-param: Any?
            }
            """;

        Assertions.assertEquals(expectedModuleContent, moduleContent);
    }

    @Test
    void buildModuleContentWithNoFunctionsAndClassesReturnsFormattedModuleContent() {
        // given
        AnnotatedPythonModule testModule = new AnnotatedPythonModule(
            "test-module",
            Collections.emptyList(),
            Collections.emptyList(),
            Collections.emptyList(),
            Collections.emptyList(),
            Collections.emptyList()
        );

        // when
        ModuleStubContentBuilder moduleStubContentBuilder =
            new ModuleStubContentBuilder(testModule);
        String moduleContent = moduleStubContentBuilder.buildModuleContent();

        //then
        String expectedModuleContent = """
            package simpleml.test-module
            """;

        Assertions.assertEquals(expectedModuleContent, moduleContent);
    }
}

File: server/src/test/kotlin/com/larsreimann/api_editor/codegen/ModuleStubContentBuilderTest.kt
package com.larsreimann.api_editor.codegen

import com.larsreimann.api_editor.model.AnnotatedPythonClass
import com.larsreimann.api_editor.model.AnnotatedPythonFunction
import com.larsreimann.api_editor.model.AnnotatedPythonModule
import com.larsreimann.api_editor.model.AnnotatedPythonParameter
import com.larsreimann.api_editor.model.AnnotatedPythonResult
import com.larsreimann.api_editor.model.PythonFromImport
import com.larsreimann.api_editor.model.PythonImport
import com.larsreimann.api_editor.model.PythonParameterAssignment
import de.unibonn.simpleml.SimpleMLStandaloneSetup
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class ModuleStubContentBuilderTest {

    @BeforeEach
    fun initSimpleML() {
        SimpleMLStandaloneSetup.doSetup()
    }

    @Test
    fun buildModuleContentReturnsFormattedModuleContent() {
        // given
        val testClass = AnnotatedPythonClass(
            "testClass",
            "testModule.testClass",
            listOf("test-decorator"),
            listOf("test-superclass"),
            listOf(
                AnnotatedPythonFunction(
                    "testClassFunction",
                    "testModule.testClass.testClassFunction",
                    listOf("decorators"),
                    listOf(
                        AnnotatedPythonParameter(
                            "onlyParam",
                            "testModule.testClass.testClassFunction.onlyParam",
                            "'defaultValue'",
                            PythonParameterAssignment.POSITION_OR_NAME,
                            true,
                            "typeInDocs",
                            "description",
                            mutableListOf()
                        )
                    ),
                    emptyList(),
                    true,
                    "description",
                    "fullDocstring",
                    mutableListOf()
                ),
                AnnotatedPythonFunction(
                    "__init__",
                    "testModule.testClass.__init__",
                    listOf("decorators"),
                    listOf(
                        AnnotatedPythonParameter(
                            "onlyParam",
                            "testModule.testClass.__init__.onlyParam",
                            "'defaultValue'",
                            PythonParameterAssignment.POSITION_OR_NAME,
                            true,
                            "typeInDocs",
                            "description", mutableListOf()
                        )
                    ),
                    emptyList(),
                    true,
                    "description",
                    "fullDocstring", mutableListOf()
                )
            ),
            "Lorem ipsum",
            "Lorem ipsum", mutableListOf()
        )
        val testModule = AnnotatedPythonModule(
            "testModule", emptyList(), emptyList(),
            listOf(
                testClass
            ),
            listOf(
                AnnotatedPythonFunction(
                    "function_module_1",
                    "test.module_1.function_module_1",
                    listOf("test-decorator"),
                    listOf(
                        AnnotatedPythonParameter(
                            "param1",
                            "test.module_1.function_module_1.param1",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum", mutableListOf()
                        ),
                        AnnotatedPythonParameter(
                            "param2",
                            "test.module_1.function_module_1.param2",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum", mutableListOf()
                        ),
                        AnnotatedPythonParameter(
                            "param3",
                            "test.module_1.function_module_1.param3",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum", mutableListOf()
                        )
                    ),
                    listOf(
                        AnnotatedPythonResult(
                            "testResult",
                            "str",
                            "str",
                            "Lorem ipsum", mutableListOf()
                        )
                    ),
                    true,
                    "Lorem ipsum",
                    "Lorem ipsum", mutableListOf()
                ),
                AnnotatedPythonFunction(
                    "testFunction",
                    "testModule.testFunction",
                    listOf("test-decorator"),
                    listOf(
                        AnnotatedPythonParameter(
                            "testParameter",
                            "testModule.testFunction.testParameter",
                            "42",
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "int",
                            "Lorem ipsum", mutableListOf()
                        )
                    ),
                    listOf(
                        AnnotatedPythonResult(
                            "testResult",
                            "str",
                            "str",
                            "Lorem ipsum", mutableListOf()
                        )
                    ),
                    true,
                    "Lorem ipsum",
                    "Lorem ipsum", mutableListOf()
                )
            ),
            mutableListOf()
        )

        // when
        val moduleStubContentBuilder = ModuleStubContentBuilder(testModule)
        val moduleContent = moduleStubContentBuilder.buildModuleContent()

        // then
        val expectedModuleContent: String = """
            |package simpleml.testModule
            |
            |class testClass(onlyParam: Any? or "defaultValue") {
            |    attr onlyParam: Any?
            |
            |    fun testClassFunction(onlyParam: Any? or "defaultValue")
            |}
            |
            |fun function_module_1(param1: String, param2: String, param3: String) -> testResult: String
            |
            |fun testFunction(testParameter: Int or 42) -> testResult: String
            |""".trimMargin()
        Assertions.assertEquals(expectedModuleContent, moduleContent)
    }

    @Test
    fun buildModuleContentWithNoClassesReturnsFormattedModuleContent() {
        // given
        val testModule = AnnotatedPythonModule(
            "testModule", emptyList(), emptyList(), emptyList(),
            listOf(
                AnnotatedPythonFunction(
                    "function_module_1",
                    "test.module_1.function_module_1",
                    listOf("test-decorator"),
                    listOf(
                        AnnotatedPythonParameter(
                            "param1",
                            "test.module_1.function_module_1.param1",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum", mutableListOf()
                        ),
                        AnnotatedPythonParameter(
                            "param2",
                            "test.module_1.function_module_1.param2",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum", mutableListOf()
                        ),
                        AnnotatedPythonParameter(
                            "param3",
                            "test.module_1.function_module_1.param3",
                            null,
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "str",
                            "Lorem ipsum", mutableListOf()
                        )
                    ),
                    listOf(
                        AnnotatedPythonResult(
                            "testResult",
                            "str",
                            "str",
                            "Lorem ipsum", mutableListOf()
                        )
                    ),
                    true,
                    "Lorem ipsum",
                    "Lorem ipsum", mutableListOf()
                ),
                AnnotatedPythonFunction(
                    "testFunction",
                    "testModule.testFunction",
                    listOf("test-decorator"),
                    listOf(
                        AnnotatedPythonParameter(
                            "testParameter",
                            "testModule.testFunction.testParameter",
                            "42",
                            PythonParameterAssignment.NAME_ONLY,
                            true,
                            "int",
                            "Lorem ipsum", mutableListOf()
                        )
                    ),
                    listOf(
                        AnnotatedPythonResult(
                            "testResult",
                            "str",
                            "str",
                            "Lorem ipsum", mutableListOf()
                        )
                    ),
                    true,
                    "Lorem ipsum",
                    "Lorem ipsum", mutableListOf()
                )
            ),
            mutableListOf()
        )

        // when
        val moduleStubContentBuilder = ModuleStubContentBuilder(testModule)
        val moduleContent = moduleStubContentBuilder.buildModuleContent()

        // then
        val expectedModuleContent: String = """
            |package simpleml.testModule
            |
            |fun function_module_1(param1: String, param2: String, param3: String) -> testResult: String
            |
            |fun testFunction(testParameter: Int or 42) -> testResult: String
            |""".trimMargin()
        Assertions.assertEquals(expectedModuleContent, moduleContent)
    }

    @Test
    fun buildModuleContentWithOnlyConstructorReturnsFormattedModuleContent() {
        // given
        val testClass = AnnotatedPythonClass(
            "testClass",
            "testModule.testClass",
            listOf("test-decorator"),
            listOf("test-superclass"),
            listOf(
                AnnotatedPythonFunction(
                    "__init__",
                    "testModule.testClass.__init__",
                    listOf("decorators"),
                    listOf(
                        AnnotatedPythonParameter(
                            "onlyParam",
                            "testModule.testClass.__init__.onlyParam",
                            "'defaultValue'",
                            PythonParameterAssignment.POSITION_OR_NAME,
                            true,
                            "typeInDocs",
                            "description", mutableListOf()
                        )
                    ),
                    emptyList(),
                    true,
                    "description",
                    "fullDocstring", mutableListOf()
                )
            ),
            "Lorem ipsum",
            "Lorem ipsum", mutableListOf()
        )
        val testModule = AnnotatedPythonModule(
            "testModule",
            listOf(
                PythonImport(
                    "test-import1",
                    "test-alias"
                )
            ),
            listOf(
                PythonFromImport(
                    "test-from-import1",
                    "test-declaration1",
                    null
                )
            ),
            listOf(
                testClass
            ),
            emptyList(), mutableListOf()
        )

        // when
        val moduleStubContentBuilder = ModuleStubContentBuilder(testModule)
        val moduleContent = moduleStubContentBuilder.buildModuleContent()

        // then
        val expectedModuleContent: String = """
            |package simpleml.testModule
            |
            |class testClass(onlyParam: Any? or "defaultValue") {
            |    attr onlyParam: Any?
            |}
            |""".trimMargin()
        Assertions.assertEquals(expectedModuleContent, moduleContent)
    }

    @Test
    fun buildModuleContentWithNoFunctionsAndClassesReturnsFormattedModuleContent() {
        // given
        val testModule = AnnotatedPythonModule(
            "testModule", emptyList(), emptyList(), emptyList(), emptyList(), mutableListOf()
        )

        // when
        val moduleStubContentBuilder = ModuleStubContentBuilder(testModule)
        val moduleContent = moduleStubContentBuilder.buildModuleContent()

        // then
        val expectedModuleContent: String = """
            |package simpleml.testModule
            |""".trimMargin()
        Assertions.assertEquals(expectedModuleContent, moduleContent)
    }
}
